Transform <Badge>, <Breadcrumb>, <BreadcrumbLink>, <CardCarouselTitle> and <CardCarouselSubtitle> to typescript

diana-villalvazo-wgu opened the pull request:
## Description

As part of the effort to "Convert components to TypeScript" https://github.com/openedx/paragon/issues/3739
Components converted to typescript
- `<Badge>`
- `<Breadcrumb>`
- `<BreadcrumbLink>`
- `<CardCarouselTitle>`
- `<CardCarouselSubtitle>`

### Deploy Preview

- `<Badge>`- https://deploy-preview-3807--paragon-openedx-v23.netlify.app/components/badge
- `<Breadcrumb>` - https://deploy-preview-3807--paragon-openedx-v23.netlify.app/components/breadcrumb
- `<BreadcrumbLink>`
- `<CardCarouselTitle>` 
- `<CardCarouselSubtitle>`

## Merge Checklist

* [x] If your update includes visual changes, have they been reviewed by a designer? Send them a link to the Netlify deploy preview, if applicable.
* [x] Does your change adhere to the documented [style conventions](https://github.com/openedx/paragon/blob/master/docs/decisions/0012-css-styling-conventions)?
* [x] Do any prop types have missing descriptions in the Props API tables in the documentation site (check deploy preview)?
* [x] Were your changes tested using all available themes (see theme switcher in the header of the deploy preview, under the "Settings" icon)?
* [ ] Were your changes tested in the `example` app?
* [x] Is there adequate test coverage for your changes?
* [ ] Consider whether this change needs to reviewed/QA'ed for accessibility (a11y). If so, please request an a11y review for the PR in the [#wg-paragon](https://openedx.slack.com/archives/C02NR285KV4) Open edX Slack channel.

## Post-merge Checklist

* [ ] Verify your changes were released to [NPM](https://www.npmjs.com/package/@openedx/paragon) at the expected version.
* [ ] If you'd like, [share](https://github.com/openedx/paragon/discussions/new?category=show-and-tell) your contribution in [#show-and-tell](https://github.com/openedx/paragon/discussions/categories/show-and-tell).
* [ ] 🎉 🙌 Celebrate! Thanks for your contribution.

openedx-webhooks commented:
Thanks for the pull request, @diana-villalvazo-wgu!


This repository is currently maintained by `@openedx/committers-paragon`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

diana-villalvazo-wgu committed:
refactor: transform components into Typescript

diana-villalvazo-wgu committed:
docs: improve variant doc

navinkarkera commented on src/Breadcrumb/BreadcrumbLink.tsx:
```diff
@@ -1,15 +1,30 @@
+  as: React.ElementType;
+  clickHandler?: (event: React.MouseEvent, link: any) => void;
+  linkProps: {
+    label: string;
```
Thanks for these changes!

Could we accept a `ReactNode` here instead of string, so that we can add icons or update the design of the label text.

Example: 

<img width="419" height="93" alt="Image" src="https://github.com/user-attachments/assets/3e2dcab3-be9f-46e0-8a3b-9a42a32f291e" />

brian-smith-tcril replied:
That is a change that would need to be discussed and should not "sneak in" as part of a PR converting `.jsx`/`PropTypes` to `.tsx`/`TypeScript` types.

navinkarkera replied:
Sure, I don't want to "sneak in" any changes :+1:

bradenmacdonald commented on src/Badge/index.tsx:
```diff
@@ -0,0 +1,32 @@
+  bsPrefix?: string;
+}
+
+const Badge = React.forwardRef(({
```
If you temporarily put this code in the `Badge` file (or any typescript file), you'll see that this doesn't quite have the right types:

```tsx
// @ts-expect-error implicit <span> does not support 'checked'
export const testOne = <Badge checked />;
// @ts-expect-error explicit <span> does not support 'checked'
export const testTwo = <Badge as="span" checked />;
// Should be no error - input does support 'checked':
export const testThree = <Badge as="input" checked />;
```

<img width="534" height="179" alt="Screenshot 2025-10-06 at 4 55 13 PM" src="https://github.com/user-attachments/assets/18eb941e-0b62-4bc1-acd5-d0a881d3a467" />

The change in 87a0537a0c9836cd03ed49c52739e6e7ada6c8a1 should fix it though.

diana-villalvazo-wgu replied:
this change breaks some tests 😅 checking if i can fix them 👀

bradenmacdonald replied:
Oops, guess I didn't check it thoroughly enough. You can just copy what's in `Badge` without the other change I suggested to `bootstrap.ts` and it should pass the type checks. But I was seeing an error in VS Code that I'd like to figure out how to solve.

bradenmacdonald commented on src/Card/CardCarousel/CardCarouselSubtitle.tsx:
```diff
@@ -1,8 +1,16 @@
+  className?: string;
+}
+
+function CardCarouselSubtitle({ children, as, className }: CardCarouselSubtitleProps) {
```
Note: this looks good. We don't need any complicated handling of `as` here because this doesn't pass `...props` to the child component.

bradenmacdonald commented on src/Card/CardCarousel/CardCarouselSubtitle.tsx:
```diff
@@ -1,8 +1,16 @@
+  /** Specifies contents of the component. */
+  children: React.ReactNode;
+  /** Specifies the base element */
+  as?: string | React.ElementType;
```
```suggestion
  as?: React.ElementType;
```
`React.ElementType` already allows strings like `div`, `span`, etc. - or is there some other non-valid-element string we need to allow?

bradenmacdonald commented on src/Card/CardCarousel/CardCarouselTitle.tsx:
```diff
@@ -0,0 +1,19 @@
+
+interface CardCarouselTitleProps {
+  children: React.ReactNode;
+  as?: React.ElementType | string;
```
```suggestion
  as?: React.ElementType;
```

bradenmacdonald reviewed (commented):
Looks great, just want to get the typing of `as` right on `Badge` - see below. 

(I know that `Badge` is probably rarely using custom elements anyways, but we should try to get the typing right in general so it's easy to apply the same patterns across the board.)

diana-villalvazo-wgu committed:
fix: address feedback comments

diana-villalvazo-wgu committed:
fix: badge type

bradenmacdonald commented:
@diana-villalvazo-wgu Is this PR ready for review again? Sorry if I dropped the ball here. Would you mind rebasing it as well?

diana-villalvazo-wgu commented:
> @diana-villalvazo-wgu Is this PR ready for review again? Sorry if I dropped the ball here. Would you mind rebasing it as well?

yes, i addressed the feedback, and just did rebase 👍  thanks Braden

bradenmacdonald reviewed (approved):
Nice, thanks!!
